smarts/core/provider.py

```diff
@@ -76,3 +88,50 @@ def reset(self):
 
     def teardown(self):
         raise NotImplementedError
+
+    def recover(self, scenario, elapsed_sim_time: float, error: Exception) -> bool:
+        return False
+
+    @property
+    def connected(self):
+        """Determine if the provider is still responsive. (e.g. the case that the provider is
+        sending provider state over the internet and has stopped responding)
+        """
+        return True
+
+    @property
+    def recovery_options(self) -> ProviderRecoveryOptions:
+        """If this provider must respond each step for an episode to continue."""
+        return ProviderRecoveryOptions.EXPERIMENT_REQUIRED
```

Comment:
I don't think it's possible (or desirable) for the _provider itself_ to know whether or not it's "necessary" (and how) in the context of the overall simulation.  That may depend on things outside of its scope and control.

Hence, I don't think this is in the right place.

Reply:
Right, it would be more reasonable for that information to be applied at the time when the provider is added.

Reply:
I have addressed it from SMARTS level